kernel/src/sched.rs

```diff
@@ -181,6 +181,7 @@ impl Kernel {
             process_identifier_max: Cell::new(0),
             grant_counter: Cell::new(0),
             grants_finalized: Cell::new(false),
+            grant_num_mapping: [None; config::CONFIG.max_drivers],
```

Comment:
We could get rid of this check. If we do so, it makes it possible for a board to introduce UB in userspace if a capsule creates 2 grants, and is intentionally written so as to maliciously swap callbacks stored in those two grants.

I don't think there is any other potential attack vector enabled by removing this check -- @lschuermann do you agree?

I am fine with removing the check. If we do remove the check, I wonder if we should mark `create_grant()` as `unsafe` -- misusing it cannot create unsoundness in the kernel, but could lead to it in the rust userspace. I suppose the answer is still probably "no".

Reply:
This topic seems sufficiently complex to contain hidden 

> I don't think there is any other potential attack vector enabled by removing this check

I'd carefully agree with this, but the interplay of all parts involved makes this rather hard to be absolutely sure we're not creating a loophole here :smile:.

> If we do remove the check, I wonder if we should mark `create_grant()` as `unsafe` -- misusing it cannot create unsoundness in the kernel, but could lead to it in the rust userspace. I suppose the answer is still probably "no".

I don't think unsafe applies here, as it's not behavior Rust considers to be unsafe, at least not with respect to the current application (i.e. the kernel). I think the kernel giving guarantees about certain behavior to userspace is rather related to semantically correct kernel code.

Reply:
@jrvanwhy do you have any thoughts on this?

Reply:
After some more discussion with Brad, the attack vector of "one capsule being initialized with the same driver number as another capsule" still exists, if both capsules collude, and one of the capsules is never called. This is a more subtle, harder to catch error, so I am in favor of keeping the check.

Reply:
After I quick discussion with @hudson-ayers, I still think this check could be useful but
- we don't create unsoundness in the kernel by not checking this
- given I'm more worried about out-of-tree boards, assuming the in-tree boards don't assign multiple grants per driver or mess this up another way, we'd still check that there are no accidental `Upcall` swaps by having this mechanism (aka. validate in-tree capsule correctness). That doesn't hold for out-of-tree capsules though.

However, @hudson-ayers' points are valid, so in general I'm indifferent about this. I do think that we could remove this check if it turns out to hurt performance or ergonomics significantly.

Reply:
I've changed my mind, because if two capsules set `DRIVER_NUM` in their files to the same number, and a board blindly uses it, then they could collude to return bad upcalls. This is still a board error, but would be a pretty subtle error. So I think it's better to have an explicit check.